ddtrace/appsec/_api_security/api_manager.py

```diff
@@ -0,0 +1,173 @@
+metrics = Metrics(namespace="datadog.api_security")
+_sentinel = object()
+
+
+class TooLargeSchemaException(Exception):
+    pass
+
+
+class APIManager(Service):
+    COLLECTED = [
+        (SPAN_DATA_NAMES.REQUEST_HEADERS_NO_COOKIES, "_dd.appsec.s.req.headers", dict),
+        (SPAN_DATA_NAMES.REQUEST_QUERY, "_dd.appsec.s.req.query", dict),
+        (SPAN_DATA_NAMES.REQUEST_PATH_PARAMS, "_dd.appsec.s.req.params", dict),
+        (SPAN_DATA_NAMES.REQUEST_BODY, "_dd.appsec.s.req.body", None),
+        (SPAN_DATA_NAMES.RESPONSE_HEADERS_NO_COOKIES, "_dd.appsec.s.res.headers", dict),
+        (SPAN_DATA_NAMES.RESPONSE_BODY, "_dd.appsec.s.res.body", None),
+    ]
+    GLOBAL_RATE_LIMIT = 50.0  # requests per seconds
+
+    INTERVAL_PER_ROUTE = 15.0  # seconds between requests
+
+    _instance = None  # type: Optional[APIManager]
+
+    @classmethod
+    def enable(cls):
+        # type: () -> None
+        if cls._instance is not None:
+            log.info("%s already enabled", cls.__name__)
+            return
+
+        log.info("Enabling %s", cls.__name__)
+        metrics.enable()
+        cls._instance = cls()
+        cls._instance.start()
+        log.info("%s enabled", cls.__name__)
+
+    @classmethod
+    def disable(cls):
+        # type: () -> None
+        if cls._instance is None:
+            log.info("%s not enabled", cls.__name__)
+            return
+
+        log.info("Disabling %s", cls.__name__)
+        cls._instance.stop()
+        cls._instance = None
+        metrics.disable()
+        log.info("%s disabled", cls.__name__)
+
+    def __init__(self):
+        # type: () -> None
+        super(APIManager, self).__init__()
+        try:
+            self.INTERVAL_PER_ROUTE = float(os.environ.get("_DD_API_SECURITY_INTERVAL_PER_ROUTE", "15.0"))
+        except BaseException:
+            pass
+
+        self._schema_meter = metrics.get_meter("schema")
+        self._log_limiter = RateLimiter(limit_rate=1.0, raise_on_exceed=False)
+        self._global_rate_limiter = RateLimiter(
+            limit_rate=self.GLOBAL_RATE_LIMIT,
+            raise_on_exceed=False,
+        )
+        self._rate_limiter_by_route = dict()  # type: Dict[str, RateLimiter]
+
+        log.info("%s initialized", self.__class__.__name__)
+
+    def _stop_service(self):
+        # type: () -> None
+        remove_context_callback(self._schema_callback, global_callback=True)
+
+    def _start_service(self):
+        # type: () -> None
+        add_context_callback(self._schema_callback, global_callback=True)
+
+    def _should_collect_schema(self, env):
+        if self.INTERVAL_PER_ROUTE <= 0.0:
+            return True
+        method = env.waf_addresses.get(SPAN_DATA_NAMES.REQUEST_METHOD)
+        route = env.waf_addresses.get(SPAN_DATA_NAMES.REQUEST_ROUTE)
+        # Framework is not fully supported
+        if not method or not route:
+            log.debug("unsupported groupkey for api security [method %s] [route %s]", bool(method), bool(route))
```

Comment:
this log is on the request path, it's probably better to use `self._log_limiter`